File: _legacy_archive_20260806/import-renaissance-baroque-to-chromadb.py
#!/usr/bin/env python3
"""
匯入Renaissance和Baroque資料到ChromaDB向量資料庫
支援所有向量基礎的RAG方法檢索
"""

import json
import chromadb
from chromadb.config import Settings
import sys
from datetime import datetime

class ChromaDBImporter:
    def __init__(self):
        # 連接到ChromaDB (使用Docker容器中的ChromaDB)
        self.client = chromadb.HttpClient(
            host='localhost',
            port=8000,
            settings=Settings(allow_reset=True, anonymized_telemetry=False)
        )

        # 創建或獲取collection
        self.collection_name = "art_history"
        try:
            self.collection = self.client.get_collection(self.collection_name)
            print(f"✅ 已連接到現有collection: {self.collection_name}")
        except:
            self.collection = self.client.create_collection(
                name=self.collection_name,
                metadata={"description": "Art history knowledge base with Renaissance and Baroque artworks"}
            )
            print(f"✅ 已創建新collection: {self.collection_name}")

        self.stats = {
            'total': 0,
            'renaissance': 0,
            'baroque': 0,
            'other': 0,
            'with_images': 0
        }

    def create_artwork_text(self, artwork):
        """
        創建藝術作品的文本表示用於嵌入
        """
        parts = []

        # 基本資訊
        if artwork.get('title'):
            parts.append(f"Title: {artwork['title']}")

        if artwork.get('artist'):
            parts.append(f"Artist: {artwork['artist']}")

        if artwork.get('date'):
            parts.append(f"Date: {artwork['date']}")

        if artwork.get('period'):
            parts.append(f"Period: {artwork['period']}")

        # 媒材和分類
        if artwork.get('medium'):
            parts.append(f"Medium: {artwork['medium']}")

        if artwork.get('classification'):
            parts.append(f"Classification: {artwork['classification']}")

        # 博物館資訊
        if artwork.get('museum'):
            parts.append(f"Museum: {artwork['museum']}")

        if artwork.get('department'):
            parts.append(f"Department: {artwork['department']}")

        # 文化和地理資訊
        if artwork.get('culture'):
            parts.append(f"Culture: {artwork['culture']}")

        if artwork.get('geography'):
            parts.append(f"Geography: {artwork['geography']}")

        return " | ".join(parts)

    def import_artworks(self, json_file_path):
        """
        從JSON檔案匯入藝術作品
        """
        print(f"\n📖 讀取資料檔案: {json_file_path}")

        with open(json_file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)

        artworks = data.get('artworks', [])
        print(f"✅ 找到 {len(artworks)} 件藝術作品")

        # 準備批次資料
        documents = []
        metadatas = []
        ids = []

        for artwork in artworks:
            # 創建文本表示
            text = self.create_artwork_text(artwork)

            # 創建metadata
            metadata = {
                'met_id': str(artwork.get('id', '')),
                'title': artwork.get('title', 'Unknown')[:500],  # ChromaDB限制長度
                'artist': artwork.get('artist', 'Unknown')[:200],
                'date': artwork.get('date', '')[:100],
                'period': artwork.get('period', '')[:50],
                'medium': artwork.get('medium', '')[:200],
                'classification': artwork.get('classification', '')[:100],
                'department': artwork.get('department', '')[:100],
                'culture': artwork.get('culture', '')[:100],
                'is_highlight': str(artwork.get('is_highlight', False)),
                'has_image': str(bool(artwork.get('image'))),
                'url': artwork.get('url', '')[:500]
            }

            # 添加到批次
            documents.append(text)
            metadatas.append(metadata)
            ids.append(f"met_{artwork.get('id', len(ids))}")

            # 更新統計
            self.stats['total'] += 1
            period = artwork.get('period', '').lower()
            if 'renaissance' in period:
                self.stats['renaissance'] += 1
            elif 'baroque' in period:
                self.stats['baroque'] += 1
            else:
                self.stats['other'] += 1

            if artwork.get('image'):
                self.stats['with_images'] += 1

        # 批次匯入到ChromaDB
        print(f"\n🔄 開始匯入 {len(documents)} 件作品到ChromaDB...")

        # ChromaDB限制每次最多41666個文檔，我們分批處理
        batch_size = 1000
        for i in range(0, len(documents), batch_size):
            batch_docs = documents[i:i+batch_size]
            batch_metas = metadatas[i:i+batch_size]
            batch_ids = ids[i:i+batch_size]

            try:
                self.collection.add(
                    documents=batch_docs,
                    metadatas=batch_metas,
                    ids=batch_ids
                )
                print(f"  ✅ 已匯入 {i+len(batch_docs)}/{len(documents)} 件作品")
            except Exception as e:
                print(f"  ⚠️ 批次 {i}-{i+batch_size} 匯入失敗: {e}")
                # 嘗試逐個匯入以找出問題
                for j, (doc, meta, doc_id) in enumerate(zip(batch_docs, batch_metas, batch_ids)):
                    try:
                        self.collection.add(
                            documents=[doc],
                            metadatas=[meta],
                            ids=[doc_id]
                        )
                    except Exception as e2:
                        print(f"    ❌ 作品 {doc_id} 匯入失敗: {e2}")

        print(f"\n✅ ChromaDB匯入完成！")
        return True

    def verify_import(self):
        """
        驗證匯入結果
        """
        print(f"\n╔═══════════════════════════════════════════════════════════╗")
        print(f"║            ChromaDB 匯入驗證                              ║")
        print(f"╚═══════════════════════════════════════════════════════════╝")

        # 獲取collection資訊
        count = self.collection.count()
        print(f"\n📊 Collection統計:")
        print(f"   Collection名稱: {self.collection_name}")
        print(f"   總文檔數: {count}")

        print(f"\n📊 匯入統計:")
        print(f"   總作品數: {self.stats['total']}")
        print(f"   Renaissance: {self.stats['renaissance']}")
        print(f"   Baroque: {self.stats['baroque']}")
        print(f"   其他相關: {self.stats['other']}")
        print(f"   含圖片: {self.stats['with_images']}")

        # 測試查詢
        print(f"\n🔍 測試向量檢索:")

        test_queries = [
            "Leonardo da Vinci",
            "Renaissance paintings",
            "Baroque art",
            "Rembrandt"
        ]

        for query in test_queries:
            try:
                results = self.collection.query(
                    query_texts=[query],
                    n_results=3
                )

                if results and results['ids'] and len(results['ids'][0]) > 0:
                    print(f"\n   ✅ 查詢 '{query}': 找到 {len(results['ids'][0])} 個結果")
                    # 顯示第一個結果
                    if results['metadatas'] and len(results['metadatas'][0]) > 0:
                        first = results['metadatas'][0][0]
                        print(f"      - {first.get('title', 'Unknown')}")
                        print(f"        藝術家: {first.get('artist', 'Unknown')}")
                        print(f"        時期: {first.get('period', 'Unknown')}")
                else:
                    print(f"   ⚠️ 查詢 '{query}': 無結果")
            except Exception as e:
                print(f"   ❌ 查詢 '{query}' 失敗: {e}")

        print(f"\n✅ ChromaDB驗證完成！")

        return count > 0

    def print_summary(self):
        """
        打印最終摘要
        """
        print(f"\n╔═══════════════════════════════════════════════════════════╗")
        print(f"║            ChromaDB 匯入成功                              ║")
        print(f"╚═══════════════════════════════════════════════════════════╝")

        print(f"\n🎉 Renaissance和Baroque資料已成功匯入ChromaDB！")
        print(f"\n📊 現在支援的RAG方法:")
        print(f"   ✅ VECTOR_ONLY - 純向量檢索")
        print(f"   ✅ GRAPH_ONLY - 純圖譜檢索")
        print(f"   ✅ HYBRID_BALANCED - 混合檢索")
        print(f"   ✅ ADAPTIVE - 自適應檢索")
        print(f"   ✅ SPECIALIZED - 專門檢索")
        print(f"   ✅ ADVANCED_RAG - 進階RAG")
        print(f"   ✅ SELF_RAG - 自我RAG")
        print(f"   ✅ AGENTIC_RAG - 代理RAG")
        print(f"   ✅ NAIVE_RAG - 簡單RAG")

        print(f"\n🌐 ChromaDB訪問:")
        print(f"   URL: http://localhost:8000")
        print(f"   Collection: {self.collection_name}")
        print(f"   文檔數: {self.stats['total']}")

        print(f"\n🚀 下一步:")
        print(f"   1. 測試所有RAG策略的檢索功能")
        print(f"   2. 驗證OpenWebUI可使用所有RAG方法")
        print(f"   3. 生成完整驗證報告")

def main():
    print("╔═══════════════════════════════════════════════════════════╗")
    print("║    Renaissance & Baroque 資料匯入 ChromaDB               ║")
    print("║    支援所有向量基礎的RAG方法                              ║")
    print("╚═══════════════════════════════════════════════════════════╝")

    # 資料檔案路徑
    data_file = "/mnt/c/Users/ssking1999/Desktop/藝術史資料庫/art-history-database/data/raw/renaissance_baroque_2025-10-15T11-42-24-847Z.json"

    try:
        # 創建匯入器
        importer = ChromaDBImporter()

        # 匯入資料
        success = importer.import_artworks(data_file)

        if success:
            # 驗證匯入
            importer.verify_import()

            # 打印摘要
            importer.print_summary()

            print(f"\n✅ 所有操作完成！")
            return 0
        else:
            print(f"\n❌ 匯入失敗")
            return 1

    except Exception as e:
        print(f"\n❌ 錯誤: {e}")
        import traceback
        traceback.print_exc()
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: _legacy_archive_20260806/import-renaissance-baroque-to-chromadb.js
#!/usr/bin/env node
/**
 * 將Renaissance和Baroque資料匯入ChromaDB向量資料庫
 */

const { ChromaClient } = require('chromadb');
const fs = require('fs');
const path = require('path');

// ChromaDB設定
const CHROMA_URL = process.env.CHROMA_URL || 'http://localhost:8001';
const COLLECTION_NAME = 'art_history_renaissance_baroque';

class ChromaDBImporter {
    constructor() {
        this.client = new ChromaClient({ path: CHROMA_URL });
        this.collection = null;
        this.stats = {
            imported: 0,
            skipped: 0,
            errors: 0
        };
    }

    async connect() {
        try {
            // 測試連接
            await this.client.heartbeat();
            console.log('✅ ChromaDB連接成功');

            // 獲取或創建集合
            try {
                this.collection = await this.client.getOrCreateCollection({
                    name: COLLECTION_NAME,
                    metadata: {
                        description: 'Renaissance and Baroque period artworks',
                        source: 'Metropolitan Museum of Art',
                        created_at: new Date().toISOString()
                    }
                });
                console.log(`✅ 使用集合: ${COLLECTION_NAME}`);
            } catch (error) {
                console.log(`創建新集合: ${COLLECTION_NAME}`);
                this.collection = await this.client.createCollection({
                    name: COLLECTION_NAME,
                    metadata: {
                        description: 'Renaissance and Baroque period artworks',
                        source: 'Metropolitan Museum of Art',
                        created_at: new Date().toISOString()
                    }
                });
            }

            return true;
        } catch (error) {
            console.error('❌ ChromaDB連接失敗:', error.message);
            return false;
        }
    }

    async importRenaissanceBaroqueData(filePath) {
        console.log(`\n🔄 開始匯入: ${path.basename(filePath)}`);

        // 讀取資料
        const rawData = fs.readFileSync(filePath, 'utf-8');
        const artworks = JSON.parse(rawData);

        console.log(`📊 總共 ${artworks.length} 件作品`);

        // 批次處理（每批100件）
        const batchSize = 100;
        for (let i = 0; i < artworks.length; i += batchSize) {
            const batch = artworks.slice(i, Math.min(i + batchSize, artworks.length));
            await this.importBatch(batch, i);

            console.log(`   ✓ 已處理 ${Math.min(i + batchSize, artworks.length)}/${artworks.length} 件作品`);
        }

        console.log(`\n✅ 匯入完成！`);
        console.log(`📈 統計:`);
        console.log(`   - 成功匯入: ${this.stats.imported}`);
        console.log(`   - 跳過: ${this.stats.skipped}`);
        console.log(`   - 錯誤: ${this.stats.errors}`);

        // 顯示集合統計
        const count = await this.collection.count();
        console.log(`\n📊 ChromaDB集合統計:`);
        console.log(`   ${COLLECTION_NAME}: ${count} 項`);
    }

    async importBatch(artworks, startIndex) {
        const ids = [];
        const documents = [];
        const metadatas = [];

        for (const artwork of artworks) {
            try {
                // 生成唯一ID
                const id = `met_${artwork.id}_${Date.now()}_${startIndex}`;

                // 創建文本文檔（用於向量化）
                const document = this.createDocument(artwork);

                // 創建元資料
                const metadata = this.createMetadata(artwork);

                ids.push(id);
                documents.push(document);
                metadatas.push(metadata);

                this.stats.imported++;

            } catch (error) {
                this.stats.errors++;
                console.error(`   ⚠️ 處理作品失敗 (ID: ${artwork.id}):`, error.message);
            }
        }

        // 批次加入到ChromaDB
        if (ids.length > 0) {
            try {
                await this.collection.add({
                    ids: ids,
                    documents: documents,
                    metadatas: metadatas
                });
            } catch (error) {
                console.error(`   ❌ 批次匯入失敗:`, error.message);
                this.stats.errors += ids.length;
            }
        }
    }

    createDocument(artwork) {
        // 創建豐富的文本描述以生成更好的向量嵌入
        const parts = [];

        // 標題
        if (artwork.title) {
            parts.push(`Title: ${artwork.title}`);
        }

        // 藝術家
        if (artwork.artist && artwork.artist !== 'Unknown Artist') {
            parts.push(`Artist: ${artwork.artist}`);
            if (artwork.artistNationality) {
                parts.push(`Artist Nationality: ${artwork.artistNationality}`);
            }
        }

        // 時期
        if (artwork.period) {
            parts.push(`Period: ${artwork.period}`);
        }

        // 日期
        if (artwork.date) {
            parts.push(`Date: ${artwork.date}`);
        }

        // 媒材
        if (artwork.medium) {
            parts.push(`Medium: ${artwork.medium}`);
        }

        // 尺寸
        if (artwork.dimensions) {
            parts.push(`Dimensions: ${artwork.dimensions}`);
        }

        // 文化
        if (artwork.culture) {
            parts.push(`Culture: ${artwork.culture}`);
        }

        // 部門
        if (artwork.department) {
            parts.push(`Department: ${artwork.department}`);
        }

        // 分類
        if (artwork.classification) {
            parts.push(`Classification: ${artwork.classification}`);
        }

        // 國家/地區
        if (artwork.country) {
            parts.push(`Country: ${artwork.country}`);
        }
        if (artwork.region) {
            parts.push(`Region: ${artwork.region}`);
        }

        // 標籤
        if (artwork.tags && artwork.tags.length > 0) {
            const tagTerms = artwork.tags.map(tag => tag.term || tag).filter(Boolean);
            if (tagTerms.length > 0) {
                parts.push(`Tags: ${tagTerms.join(', ')}`);
            }
        }

        return parts.join('. ');
    }

    createMetadata(artwork) {
        // 過濾掉null和undefined值
        const metadata = {};

        if (artwork.id) metadata.met_id = String(artwork.id);
        if (artwork.title) metadata.title = this.truncate(artwork.title, 500);
        if (artwork.artist) metadata.artist = this.truncate(artwork.artist, 200);
        if (artwork.date) metadata.date = this.truncate(artwork.date, 100);
        if (artwork.period) metadata.period = artwork.period;
        if (artwork.medium) metadata.medium = this.truncate(artwork.medium, 300);
        if (artwork.culture) metadata.culture = this.truncate(artwork.culture, 100);
        if (artwork.department) metadata.department = this.truncate(artwork.department, 100);
        if (artwork.classification) metadata.classification = this.truncate(artwork.classification, 100);
        if (artwork.country) metadata.country = this.truncate(artwork.country, 100);
        if (artwork.isHighlight !== undefined) metadata.is_highlight = artwork.isHighlight;
        if (artwork.primaryImage) metadata.has_image = true;
        if (artwork.objectURL) metadata.url = artwork.objectURL;

        // 添加時間範圍（用於過濾）
        if (artwork.beginDate && !isNaN(artwork.beginDate)) {
            metadata.begin_year = parseInt(artwork.beginDate);
        }
        if (artwork.endDate && !isNaN(artwork.endDate)) {
            metadata.end_year = parseInt(artwork.endDate);
        }

        metadata.source = 'Metropolitan Museum of Art';
        metadata.crawled_at = artwork.crawledAt || new Date().toISOString();

        return metadata;
    }

    truncate(str, maxLength) {
        if (!str) return '';
        str = String(str);
        return str.length > maxLength ? str.substring(0, maxLength) + '...' : str;
    }

    async testQuery(query = 'Renaissance painting') {
        console.log(`\n🔍 測試查詢: "${query}"`);

        const results = await this.collection.query({
            queryTexts: [query],
            nResults: 5
        });

        console.log(`📊 找到 ${results.ids[0].length} 項結果:`);

        for (let i = 0; i < results.ids[0].length; i++) {
            const metadata = results.metadatas[0][i];
            const distance = results.distances[0][i];

            console.log(`\n   ${i + 1}. ${metadata.title}`);
            console.log(`      藝術家: ${metadata.artist || 'Unknown'}`);
            console.log(`      時期: ${metadata.period || 'Unknown'}`);
            console.log(`      日期: ${metadata.date || 'Unknown'}`);
            console.log(`      相似度: ${(1 - distance).toFixed(4)}`);
        }
    }
}

async function main() {
    console.log('🎨 Renaissance & Baroque 資料匯入 ChromaDB');
    console.log('============================================\n');

    const importer = new ChromaDBImporter();

    // 連接ChromaDB
    const connected = await importer.connect();
    if (!connected) {
        process.exit(1);
    }

    // 找到最大的renaissance_baroque資料檔
    const dataDir = './data/raw';
    const files = fs.readdirSync(dataDir)
        .filter(f => f.startsWith('renaissance_baroque_') && f.endsWith('.json'))
        .map(f => ({
            name: f,
            size: fs.statSync(path.join(dataDir, f)).size
        }))
        .sort((a, b) => b.size - a.size);

    if (files.length === 0) {
        console.error('❌ 找不到renaissance_baroque資料檔');
        process.exit(1);
    }

    const latestFile = path.join(dataDir, files[0].name);
    console.log(`📁 使用檔案: ${files[0].name} (${(files[0].size / 1024).toFixed(2)} KB)\n`);

    try {
        // 匯入資料
        await importer.importRenaissanceBaroqueData(latestFile);

        // 測試查詢
        await importer.testQuery('Renaissance painting by Leonardo da Vinci');
        await importer.testQuery('Baroque sculpture');

        console.log('\n🎉 全部完成！');

    } catch (error) {
        console.error('\n❌ 匯入失敗:', error);
        console.error(error.stack);
    }
}

if (require.main === module) {
    main().catch(console.error);
}

module.exports = ChromaDBImporter;
